Add `--limit` option to `docker search`

yongtang opened the pull request:
**\- What I did**

This fix tries to address the issue raised in #23055. Currently `docker search` result caps at 25 and there is no way to allow getting more results (if exist).

This fix adds the flag `--limit` so that it is possible to return more results from the `docker search`.

**\- How I did it**

Additional flag `--limit` has been added to `docker search`. Related documentation has been updated.

**\- How to verify it**

Additional tests have been added to cover the changes.

**\- Description for the changelog**

Add `--limit` option to `docker search` so that `docker search` could return different number of search results.

**\- A picture of a cute animal (not mandatory but encouraged)**

This fix fixes #23055.

**NOTE: Pull request for engine-api will be created separately.**

Signed-off-by: Yong Tang yong.tang.github@outlook.com

thaJeztah commented:
I tested this, and it works, but we may need to add some validation / boundary checks;

Values `<= 0` are treated as "1" it seems;

```
docker search --limit=0 a
NAME      DESCRIPTION                                     STARS     OFFICIAL   AUTOMATED
ubuntu    Ubuntu is a Debian-based Linux operating s...   4008      [OK]

docker search --limit=-1 a
NAME      DESCRIPTION                                     STARS     OFFICIAL   AUTOMATED
ubuntu    Ubuntu is a Debian-based Linux operating s...   4008      [OK]
```

The maximum limit (I suspect, as set by Docker Hub), is 100, any value over that still returns 100 results.

I think we should at least document that, perhaps add an error message for values `< 1` or `> 100`

thaJeztah commented:
Design looks good to me though, so I'm moving to code review

yongtang commented:
Thanks @thaJeztah. The pull request has been updated with value check between `[1, 100]`. The documentation has been updated as well. Please let me know if there are any issues.

vdemeester commented:
@yongtang you can make a PR to `engine-api` too now (design-review passed :+1:)

yongtang commented:
Thanks @vdemeester!
I created a pull request in engine-api:
https://github.com/docker/engine-api/pull/250
Let me know if there are any issues.

thaJeztah commented:
@yongtang It's merged 👍

yongtang commented:
Thanks for the help @thaJeztah @vdemeester!
It seems that both this pull request (#23107) and another pull request (#23090) have updated engine-api and there are some overlapping conflict.
I am wondering if it is possible to review #23090 first? After #23090 is merged it will be much easier to update this pull request ( #23107 )

thaJeztah commented:
@yongtang just reviewed that PR; had one minor nit, but looks ok to go after that

yongtang committed:
Add `--limit` option to `docker search`

This fix tries to address the issue raised in #23055.
Currently `docker search` result caps at 25 and there is
no way to allow getting more results (if exist).

This fix adds the flag `--limit` so that it is possible
to return more results from the `docker search`.

Related documentation has been updated.

Additional tests have been added to cover the changes.

This fix fixes #23055.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @thaJeztah @vdemeester for the review.

The pull request has been rebase to reflect the changes in engine-api (PR #23208). Please let me know if there are any issues.

vdemeester commented:
LGTM 🐮

thaJeztah commented:
LGTM, thanks!

vdemeester commented:
docs LGTM :angel:

thaJeztah commented:
oh, and ping @albers, @sdurrheimer (for the new flag)
